### ISSUE TITLE
How to implement a DatabaseAdapter that proxy WebView calls to React Native?

### ISSUE THREAD

**Issue by KennanChan** [Nozbe/WatermelonDB#1611]
I use watermelondb in an web application(using IndexedDB as database engine) which runs inside a react-native application through webview. 

The business data increased so much that it was hard for browser to hold these data efficiently and safely so I decided to persist database into react-native layer while keeping most of the business code in the browser unchanged.

My idea is to create a `SQLiteAdapter` in react-native layer and create a `ProxySQLiteAdapter` in the browser. The `ProxySQLiteAdapter` acts as a bridge that proxies `DatabaseAdapter` calls from the browser to react-native through webview messaging, and the message is then transformed to a corresponding `SQLiteAdapter` method call. Once the `SQLiteAdapter` gets the result, it pass the result back to browser through javascript callbacks.

I created a sequence diagram to show my idea clear.
![ProxySQLiteAdapter](https://github.com/Nozbe/WatermelonDB/assets/15809890/55b2aa2a-e146-439a-8a33-0865c1f86a80)

This way I need to serialize/deserialize the arguments and return values of all the methods defined in `DatabaseAdapter` interface.

Problem is the argument of `DatabaseAdapter.query()` and `DatabaseAdapter.count()` is of type `Query<Model>` which is not plain object and I found it hard to be serialized.

Can someone give me some guide about how to serialize/deserialize a query?
Or, are there any other solutions?

**Comment by KennanChan** [comment]
I realized that I might need to serialize `Query<Model>` based on specific `DatabaseAdapter.query` implementation.
I'll go through the source code of `SQLiteDatabaseAdapter` to find out a way to serialize and deserialize.

**Comment by KennanChan** [comment]
I found that the `DatabaseAdapter` interface has been changed in v0.24.1-7 and the argument of `DatabaseAdapter.query` is now `SerializedQuery`. I'll upgrade to the latest version.

**Comment by radex** [comment]
@KennanChan As a hint, take a look at LokiJSAdapter. It has a (deprecated, unmaintained) mode in which the database runs in a web worker, so effectively a completely separate JS script with the only communications being serialized. This is equivalent to what you're trying to achieve with SQLiteAdapter, and hopefully gives you clues as for how to approach this.

If you need more assistance with this, please email me

**Comment by KennanChan** [comment]
> @KennanChan As a hint, take a look at LokiJSAdapter. It has a (deprecated, unmaintained) mode in which the database runs in a web worker, so effectively a completely separate JS script with the only communications being serialized. This is equivalent to what you're trying to achieve with SQLiteAdapter, and hopefully gives you clues as for how to approach this.
> 
> 
> 
> If you need more assistance with this, please email me

Great!
The latest version of DatabaseAdapter interface looks quite friendly to serialization. I am trying a solution based on it. 
I'll take a look at the LokiJsAdapter implementation.

**Comment by KennanChan** [comment]
I struggled to make it work based on the latest stable version of wdb!

But I still got another problem.

Wdb database layer in the webview automatically caches the records returned by the DatabaseAdapter, which may result in large memory consumption. This contributes nothing to my goal to lower the browser memory stress.

I am wondering whether the cache mode can be disabled to sacrifice some bandwidth for communication to ensure the browser memory not get exploded.

**Comment by radex** [comment]
> I am wondering whether the cache mode can be disabled

not out of the box, but you can modify it to suit your needs. At a minimum though, you need a WeakMap cache so that record references are identical

**Comment by radex** [comment]
@KennanChan Hey, curious -- did you finish or did you give up on this project?

**Comment by KennanChan** [comment]
@radex In my application, I have implemented some designs to alleviate the pressure caused by large backend databases. Firstly, I have divided the databases into smaller logical units, each containing a part of the data. This has significantly reduced memory consumption. I have also created a DatabaseManager to seamlessly switch between these smaller databases. Additionally, I have utilized an LRU cache container in the DatabaseManager to cache the most recently used databases and avoid frequent garbage collection and instantiation.

To address the issue at hand, I have made two changes. Firstly, I have sliced the database from another business perspective, resulting in more but smaller databases. This has proven to be effective in reducing pressure. Secondly, I have set the LRU cache to a smaller capacity to allow for more garbage collection. This approach has also been successful.

Although these actions have bought me more time to find a better solution, the issue is not entirely resolved as business data continues to grow.
